app/src/test/java/org/mozilla/fenix/home/BlocklistMiddlewareTest.kt

```diff
@@ -0,0 +1,288 @@
+        )
+
+        val updatedBookmark = RecentBookmark(url = "https://www.mozilla.org/")
+        store.dispatch(
+            HomeFragmentAction.Change(
+                topSites = store.state.topSites,
+                mode = store.state.mode,
+                collections = store.state.collections,
+                tip = store.state.tip,
+                showCollectionPlaceholder = store.state.showCollectionPlaceholder,
+                recentTabs = store.state.recentTabs,
+                recentBookmarks = listOf(updatedBookmark),
+                recentHistory = store.state.recentHistory
+            )
+        ).joinBlocking()
+
+        assertEquals(updatedBookmark, store.state.recentBookmarks[0])
+    }
+
+    @Test
+    fun `GIVEN non-empty blocklist with specific pages WHEN action intercepted with matching host THEN unchanged by middleware`() {
+        every { mockSettings.homescreenBlocklist } returns setOf("https://github.com/mozilla-mobile/fenix")
+
+        val middleware = BlocklistMiddleware(mockSettings)
+
+        val store = HomeFragmentStore(
+            HomeFragmentState(),
+            middlewares = listOf(middleware)
+        )
+
+        val updatedBookmark = RecentBookmark(url = "https://github.com/")
+        store.dispatch(
+            HomeFragmentAction.Change(
+                topSites = store.state.topSites,
+                mode = store.state.mode,
+                collections = store.state.collections,
+                tip = store.state.tip,
+                showCollectionPlaceholder = store.state.showCollectionPlaceholder,
+                recentTabs = store.state.recentTabs,
+                recentBookmarks = listOf(updatedBookmark),
+                recentHistory = store.state.recentHistory
+            )
+        ).joinBlocking()
+
+        assertEquals(updatedBookmark, store.state.recentBookmarks[0])
+    }
+
+    @Test
+    fun `GIVEN non-empty blocklist WHEN action intercepted with matching elements THEN filtered by middleware`() {
+        every { mockSettings.homescreenBlocklist } returns setOf("https://www.mozilla.org/")
+
+        val middleware = BlocklistMiddleware(mockSettings)
+
+        val store = HomeFragmentStore(
+            HomeFragmentState(),
+            middlewares = listOf(middleware)
+        )
+
+        val updatedBookmark = RecentBookmark(url = "https://www.mozilla.org/")
+        store.dispatch(
+            HomeFragmentAction.Change(
+                topSites = store.state.topSites,
+                mode = store.state.mode,
+                collections = store.state.collections,
+                tip = store.state.tip,
+                showCollectionPlaceholder = store.state.showCollectionPlaceholder,
+                recentTabs = store.state.recentTabs,
+                recentBookmarks = listOf(updatedBookmark),
+                recentHistory = store.state.recentHistory
+            )
+        ).joinBlocking()
+
+        assertTrue(store.state.recentBookmarks.isEmpty())
```

Comment:
```suggestion
        every { mockSettings.homescreenBlocklist } returns setOf("https://www.mozilla.org/")

        val store = HomeFragmentStore(
            HomeFragmentState(),
            middlewares = listOf(BlocklistMiddleware(mockSettings))
        )
        val updatedBookmark = RecentBookmark(url = "https://www.mozilla.org/")
        
        store.dispatch(
            HomeFragmentAction.Change(
                topSites = store.state.topSites,
                mode = store.state.mode,
                collections = store.state.collections,
                tip = store.state.tip,
                showCollectionPlaceholder = store.state.showCollectionPlaceholder,
                recentTabs = store.state.recentTabs,
                recentBookmarks = listOf(updatedBookmark),
                recentHistory = store.state.recentHistory
            )
        ).joinBlocking()

        assertTrue(store.state.recentBookmarks.isEmpty())
```

This isn't an exact science by any means, but the formatting pattern that we seem to go with in testss is that we have 'groupings':
VARIABLES INIT
<new line>
WHEN/EVERY declarations
<new line>
SOME ACTION
<new line>
VERIFY/ASSERT

Reply:
I can move the pure variable declarations up, if you want. The mock setups are required before the initialization of the middleware, otherwise `settings` won't have an answer for `private var blocklist: Set<String> = settings.homescreenBlocklist`.

Is there a document where some of these specific formatting gotchas live? JW if there's something I can point contributors (and myself) to in order to avoid bikeshedding PR cycles since I assume some of these would be hard to create lint rules for

Reply:
Only way is probably to look at a couple of existing tests file and just adopt their pattern. Usually you just want to blend in with what the project is already doing. I only called this out because of the inconsistency usage new lines between the variable declaration. The mock setup before the initialization is fine.

Reply:
I will look to write something to help formalize this. Apologizes for the bikeshedding.

Reply:
I moved variable declarations to the spot you suggested.

I definitely agree that having stylistic similarity throughout the code base makes it easier to understand any one part. IMO a "good enough" level would be something along the lines of "follows a given-when-then" format is a bigger win than specific formatting details. I just think it's important to try and keep barrier to entry as low as possible for outside contributors. That said, I do understand if we want to hold org members to a higher standard than outside contributors, and I appreciate the guidance in getting up to par.

Reply:
Quick comment regarding a project-wide style for unit tests.
Splitting the unit test in 3 sections is a common pattern - https://wiki.c2.com/?ArrangeActAssert
Instead of the proposed 4 sections we could have 3 matching exactly `arrange/act/assert` / `given/when/then` with setup including properties intializations and mocking.